packages/backend/src/invoice/service.test.ts

```diff
@@ -0,0 +1,220 @@
+      expect(invoices[20]).toBeUndefined()
+    })
+
+    test('Can change forward pagination limit', async (): Promise<void> => {
+      const pagination = {
+        first: 10
+      }
+      const invoices = await invoiceService.getUserInvoicesPage(
+        user.id,
+        pagination
+      )
+      expect(invoices).toHaveLength(10)
+      expect(invoices[0].id).toEqual(invoicesCreated[0].id)
+      expect(invoices[9].id).toEqual(invoicesCreated[9].id)
+      expect(invoices[10]).toBeUndefined()
+    })
+
+    test('Can paginate forwards from a cursor', async (): Promise<void> => {
+      const pagination = {
+        after: invoicesCreated[19].id
+      }
+      const invoices = await invoiceService.getUserInvoicesPage(
+        user.id,
+        pagination
+      )
+      expect(invoices).toHaveLength(20)
+      expect(invoices[0].id).toEqual(invoicesCreated[20].id)
+      expect(invoices[19].id).toEqual(invoicesCreated[39].id)
+      expect(invoices[20]).toBeUndefined()
+    })
+
+    test('Can paginate forwards from a cursor with a limit', async (): Promise<void> => {
+      const pagination = {
+        first: 10,
+        after: invoicesCreated[9].id
+      }
+      const invoices = await invoiceService.getUserInvoicesPage(
+        user.id,
+        pagination
+      )
+      expect(invoices).toHaveLength(10)
+      expect(invoices[0].id).toEqual(invoicesCreated[10].id)
+      expect(invoices[9].id).toEqual(invoicesCreated[19].id)
+      expect(invoices[10]).toBeUndefined()
+    })
+
+    test("Can't change backward pagination limit on it's own.", async (): Promise<void> => {
+      const pagination = {
+        last: 10
+      }
+      const invoices = invoiceService.getUserInvoicesPage(user.id, pagination)
+      await expect(invoices).rejects.toThrow(
+        "Can't paginate backwards from the start."
+      )
+    })
+
+    test('Can paginate backwards from a cursor', async (): Promise<void> => {
+      const pagination = {
+        before: invoicesCreated[20].id
+      }
+      const invoices = await invoiceService.getUserInvoicesPage(
+        user.id,
+        pagination
+      )
+      expect(invoices).toHaveLength(20)
+      expect(invoices[0].id).toEqual(invoicesCreated[0].id)
+      expect(invoices[19].id).toEqual(invoicesCreated[19].id)
+      expect(invoices[20]).toBeUndefined()
+    })
+
+    test('Can paginate backwards from a cursor with a limit', async (): Promise<void> => {
+      const pagination = {
+        last: 10,
+        before: invoicesCreated[10].id
+      }
+      const invoices = await invoiceService.getUserInvoicesPage(
+        user.id,
+        pagination
+      )
+      expect(invoices).toHaveLength(10)
+      expect(invoices[0].id).toEqual(invoicesCreated[0].id)
+      expect(invoices[9].id).toEqual(invoicesCreated[9].id)
+      expect(invoices[10]).toBeUndefined()
+    })
+
+    test('Providing before and after results in forward pagination', async (): Promise<void> => {
```

Comment:
What do you think about making `after`/`first` and `before`/`last` mutually exclusive in the `Pagination` definition?

Reply:
I'm not sure how you would achieve this? Also, `Pagination` is just mimicking the Schema input to invoices. And you wouldn't be able to change that unless you deviated from the spec.